wip: Select unit modal

sophieyang0813 opened the pull request:
wip #469 

@nbiton 1. is `/report/select-unit` route name okay? 
I am still trying to figure out why `overflow-auto` does not make the unit list scrollable. After that and searchbar implementation, this code will be ready for review 😄😄 

@franck-boullier @kiatlim 
The current implementation is that as soon as unit is selected, create report dialog pops up. This implementation saves a user one click, as a user doesn't have to click 'create report' button after selecting a unit. Is this tentative implementation okay? or 'create report' button in our current design is preferable?  

flow: select unit -> create-report page  
<img width="270" align='left' alt="screen shot 2018-09-19 at 7 11 02 pm" src="https://user-images.githubusercontent.com/25441733/45750326-2685d000-bc41-11e8-905a-74c7b5086b47.png">
 
<img width="270" align='center' alt="screen shot 2018-09-19 at 7 19 07 pm" src="https://user-images.githubusercontent.com/25441733/45750328-27b6fd00-bc41-11e8-81d2-5fc947364572.png">

* sorry, my commit messages are a bit messy and not very telling. I was going back and forth between other pull request and I just committed instead of stashing...

franck-boullier commented:
> This implementation saves a user one click, as a user doesn't have to click 'create report' button after selecting a unit. Is this tentative implementation okay?

@sophieyang0813 clever! I like your idea yes

franck-boullier commented:
@sophieyang0813 quick question: how much more complex will it be to implement that for both the 'Reports' And the 'Case' page in that PR?

The need is the same: user need to select a unit from a long list to either add or report or add a case to a unit.

franck-boullier commented:
> @sophieyang0813 clever! I like your idea yes

EDIT/change of mind 😊 : I just looked at the recommended design again.
<img width="826" alt="screen shot 2018-09-03 at 5 48 54 pm" src="https://user-images.githubusercontent.com/38183796/44980166-ab19f280-afa1-11e8-8f2d-e4a0ace8099d.png">

@sophieyang0813 

>  This implementation saves a user one click, as a user doesn't have to click 'create report' button after selecting a unit. Is this tentative implementation okay?

Problem with that idea is that if the unit does not yet exist, then the user is stuck...
I kinda like the idea that we allow the user to create new unit with just one click in case the unit does not yet exist in the list...
---> I'd recommend that we stick to the initial design actually...

nbiton commented:
@sophieyang0813 
> is /report/select-unit route name okay?

I'd try to stick to nouns for route names, so I like `/report/unit-select` better (even though it's a very subtle difference)

> I am still trying to figure out why overflow-auto does not make the unit list scrollable.

Is it still not scrollable? If it is, it might be because a container needs not only `overflow: auto;` for it to be scrollable, it also needs its height to be restricted so it doesn't grow out and overflow on **its own parent container**.
In many other layouts in Unee-t the parent container is a flex container (mostly with `flex-column`) and the scrollable child has both `flex-grow` (take up all the available space in this container not taken by other children) and `overflow-auto`. 
This is used to create responsive scrollable containers.

sophieyang0813 commented:
@nbiton Good afternoon! I am implementing the unit-select modal in the case page as per @franck-boullier's request. I faced a Bugzilla API error that did not occur when implementing in the report page... 🤔🤔Could you kindly hint me to a right direction?
<img width="611" alt="screen shot 2018-09-20 at 2 20 48 pm" src="https://user-images.githubusercontent.com/25441733/45800863-c6486a00-bce4-11e8-9ac6-cdcd93d3cd68.png">

The code, other than the case page implementation, is ready for review. The unit-select modal for the report page has been implemented. 

**Flow**
<img width="250" align='left' alt="screen shot 2018-09-20 at 2 51 12 pm" src="https://user-images.githubusercontent.com/25441733/45800846-ba5ca800-bce4-11e8-904d-51529218fa7a.png">
<img width="250" align='center' alt="screen shot 2018-09-20 at 2 51 28 pm" src="https://user-images.githubusercontent.com/25441733/45800848-bc266b80-bce4-11e8-8329-a98fa5cfd92b.png">
<img width="250"  align='left' alt="screen shot 2018-09-20 at 2 51 36 pm" src="https://user-images.githubusercontent.com/25441733/45800849-bd579880-bce4-11e8-9420-52666b71ce4d.png">
<img width="250"  align='center' alt="screen shot 2018-09-20 at 2 51 42 pm" src="https://user-images.githubusercontent.com/25441733/45800852-bf215c00-bce4-11e8-8af5-e3376893bcb5.png">

franck-boullier commented:
@sophieyang0813 app has not been deployed in Heroku ---> impossible to test

franck-boullier commented:
@sophieyang0813 
Make sure the buttons at the bottom of the modal are always visible

<img width="250"  align='left' alt="screen shot 2018-09-20 at 2 51 36 pm" src="https://user-images.githubusercontent.com/25441733/45800849-bd579880-bce4-11e8-9420-52666b71ce4d.png">

franck-boullier commented:
@sophieyang0813 
I'm assuming there should be a scroll bar on the right side of the scrollable part of the modal

<img width="250" align='left' alt="screen shot 2018-09-20 at 2 51 12 pm" src="https://user-images.githubusercontent.com/25441733/45800846-ba5ca800-bce4-11e8-904d-51529218fa7a.png">

sophieyang0813 commented:
@nbiton No worries about Bugzilla Error! It only occurs in my local environment. On Heroku, the unit-select dialog pops up when 'add the case' button is clicked. 🎉 customizing scrollbar location is still work in progress.

franck-boullier commented:
# Issue

API error when trying to select a unit with the '+' button in the '[Case](https://unee-t-pr-519.herokuapp.com/case/unit-select)' page

![image](https://user-images.githubusercontent.com/31331637/45805823-1d087080-bcf2-11e8-8d8d-57f88c5ac764.png)

franck-boullier commented:
# Issue: 

Search box disappears when user scrolls down.

Expected  Behavior: Search box is always visible 'Docked' on top of the modal

![image](https://user-images.githubusercontent.com/31331637/45805986-7ec8da80-bcf2-11e8-8933-3aa3833974ae.png)

sophieyang0813 commented:
Oh, interesting... 🤔I just checked the log. Clicking the button shows the functioning unit-select-dialog, while raising Bugzilla API error (`unit-select' is not a valid case number nor an alias to a case`)

**functioning /case/unit-select with API error** 
<img width="300" align='left' alt="screen shot 2018-09-20 at 4 29 12 pm" src="https://user-images.githubusercontent.com/25441733/45805918-593bd100-bcf2-11e8-9196-143f82776ce9.png">
<img width="300" align='center' alt="screen shot 2018-09-20 at 4 29 44 pm" src="https://user-images.githubusercontent.com/25441733/45805945-6789ed00-bcf2-11e8-8bb3-12dbad935f6f.png">

franck-boullier commented:
# Issue: 

- go to the report page
- click on the '+' icon to add a report
- Select a unit
- go to the "name your report" step
- Cancel.

## Expected result:

Go back to the 'Report' page - modal window is NOT visible

![image](https://user-images.githubusercontent.com/31331637/45806474-bdab6000-bcf3-11e8-8ff1-26a47a1f81f7.png)


## Actual result: 

Go back to the 'Report' page - modal window IS visible

![image](https://user-images.githubusercontent.com/31331637/45806436-a3718200-bcf3-11e8-82b5-b4ef8b61c4af.png)

franck-boullier commented:
Status Update:

# Fixed issues:
![image](https://user-images.githubusercontent.com/31331637/45861301-74641a80-bd9e-11e8-8475-e128c37bb80a.png)

- Action button should always be visible FIXED
- Search box disappears when user scrolls down. FIXED
- Increase margin after the button "Create Report" FIXED (this looks better than before - maybe the fact that it's docked change the perception...)
- Scroll bar is next to the list of unit (not for the whole modal) FIXED

# Pending issues:

## API error on 'Case' page:

API error when trying to select a unit with the '+' button in the '[Case](https://unee-t-pr-519.herokuapp.com/case/unit-select)' page

![image](https://user-images.githubusercontent.com/31331637/45805823-1d087080-bcf2-11e8-8d8d-57f88c5ac764.png)

## Truncated Unit names:

The names of the units are truncated if they do not fit on the screen.
It would be better if we display the full name of the unit always, even if we have to display it on several lines

![image](https://user-images.githubusercontent.com/31331637/45805539-863bb400-bcf1-11e8-8eb3-e1b5d1a4b8ab.png)

## Cursor

Is it possible to have the cursor exactly before the 'U' and not right after the magnifying icon?
![image](https://user-images.githubusercontent.com/31331637/45861098-711c5f00-bd9d-11e8-94a6-d7620aaac381.png)

## Modal stays open when user cancel report creation: 

- go to the report page
- click on the '+' icon to add a report
- Select a unit
- go to the "name your report" step
- Cancel.

### Expected result:

Go back to the 'Report' page - modal window is NOT visible

![image](https://user-images.githubusercontent.com/31331637/45806474-bdab6000-bcf3-11e8-8ff1-26a47a1f81f7.png)


### Actual result: 

Go back to the 'Report' page - modal window IS visible

![image](https://user-images.githubusercontent.com/31331637/45861301-74641a80-bd9e-11e8-8475-e128c37bb80a.png)

## Loading Icon

On the [report page](https://unee-t-pr-519.herokuapp.com/report), while the page in loading, the "loading" icon is off center for some reason

![image](https://user-images.githubusercontent.com/31331637/45805063-5e981c00-bcf0-11e8-9a77-fe0a2661c8ba.png)

## 'Square box' around the list of units is missing a bottom line:

The Box around the list of units is missing a bottom line
![image](https://user-images.githubusercontent.com/31331637/45861368-c5740e80-bd9e-11e8-97cb-53690795eb32.png)

franck-boullier commented:
Status Update:

# Fixed issues:

![image](https://user-images.githubusercontent.com/31331637/45873725-b3f22d00-bdc5-11e8-824c-b8ab3a30d4dd.png)

- Truncated Unit names: The names of the units are truncated if they do not fit on the screen. FIXED
- Loading Icon: while the page in loading, the "loading" icon is off center for some reason FIXED

# Pending issues:

## API error on 'Case' page:

API error when trying to select a unit with the '+' button in the '[Case](https://unee-t-pr-519.herokuapp.com/case/unit-select)' page

![image](https://user-images.githubusercontent.com/31331637/45805823-1d087080-bcf2-11e8-8d8d-57f88c5ac764.png)

## No need for unit descriptions:

Unit description: 
- are taking too much space 
- are not considered in the search

Best to only use the unit name  and NOT include the Unit Description
![image](https://user-images.githubusercontent.com/31331637/45874169-c6209b00-bdc6-11e8-92f9-2bc7aed38ab2.png)

## Unit names are NOT coming from the MongoDB:

The unit names MUST come from the MongoDB and NOT from the BZFE.
See https://github.com/unee-t/frontend/issues/394 for more details on this issue

## Cursor

Is it possible to have the cursor exactly before the 'U' and not right after the magnifying icon?
![image](https://user-images.githubusercontent.com/31331637/45861098-711c5f00-bd9d-11e8-94a6-d7620aaac381.png)

## Modal stays open when user cancel report creation: 

- go to the report page
- click on the '+' icon to add a report
- Select a unit
- go to the "name your report" step
- Cancel.

### Expected result:

Go back to the 'Report' page - modal window is NOT visible

![image](https://user-images.githubusercontent.com/31331637/45806474-bdab6000-bcf3-11e8-8ff1-26a47a1f81f7.png)


### Actual result: 

Go back to the 'Report' page - modal window IS visible

![image](https://user-images.githubusercontent.com/31331637/45861301-74641a80-bd9e-11e8-8475-e128c37bb80a.png)

## 'Square box' around the list of units is missing a bottom line:

The Box around the list of units is missing a bottom line
![image](https://user-images.githubusercontent.com/31331637/45861368-c5740e80-bd9e-11e8-97cb-53690795eb32.png)

sophieyang0813 commented:
@nbiton Evening! Could you kindly point how to fix Bugzilla API? This code is almost ready except couple minor issues. Feel free to comment on them  
@sebdeckers thank you for all your hints and tips this afternoon with CSS and Modal URL 🌻🌻

@franck-boullier Thank you for the status updates! Will be fixed accordingly.  
* On 'Square box' around the list of units is missing a bottom line
It is partly because material-ui/dialog only allows me to append to the scroll bar to the bod and not the container of the unit list. That customization would be much more difficult...The bottom border line can be found at the end of the list (body).
* I fixed cursor but I guess it should go slightly more to the left... 😅

nbiton commented:
> Could you kindly point how to fix Bugzilla API?

This is actually pretty tricky because of the way route matching works.
You can try adding a third route object in between the other two here: 
https://github.com/unee-t/frontend/blob/master/imports/ui/case/case-master.jsx#L41

```
{
  path: '/case/unit-select',
  RouteComp: CaseExplorer,
  exact: true
}
```

sophieyang0813 commented:
The current code is a quick fix (no more API error, and no more modal displaying after a user clicking on the back button) without `/unit-select` for modal. 

After I pushed the code, I thought we might want to have URL for modal.. 😅(do we?) So, I am incorporating modal URL again and fix the problem accordingly.

**Modal URL: work in progress**  
I am trying to figure out how to conditionally change the value of onDismissed in unit.jsx. 
Clicking 'Create Report' button dispatches directly to createReportDialog in **unit.jsx**. But createReportDialog is configured to serve report creation in the unit page.

sophieyang0813 commented:
@nbiton all the issues raised are addressed, except the case modal. The report modal works though; Report modal closes when a user cancels report creation. it has a URL (i.e. /report/unit-select). I am still figuring out how to use react DOM to debug currently rendered RouteComp?

Units with `metaData.displayName` display name from MongoDB,  while those without metaData display name from BZFE.

franck-boullier commented:
Status Update:

# Fixed issues:

- Unit descriptions should not be displayed FIXED
- Unit names should be coming from the MongoDB FIXED
- Cursor position in the search box FIXED
- If user aborts report creation, user is sent back to the list of report and modal is closed FIXED

# Pending issues:

## API error on 'Case' page:

This is a Different error from before apparently

Endless loading page when trying to select a unit with the '+' button in the '[Case](https://unee-t-pr-519.herokuapp.com/case/unit-select)' page (no error message in the console)

![image](https://user-images.githubusercontent.com/31331637/45985698-03707b80-c09a-11e8-9241-221877fec04f.png)

## Browser compatibility (FF) - Impossible to create a Report:

Test on FF on a computer.
The problem does NOT happen on Chrome on Android (Galaxy S8)

- Go to the report page
- click on the '+' icon to add a report
- Select a unit
- Click on 'Create Report'

### Expected result:

User is sent to the "name your report" step.

### Actual result:

Nothing happens, the window stays open
![image](https://user-images.githubusercontent.com/31331637/45985523-37976c80-c099-11e8-8d4a-e320a83cf0e3.png)

## 'Square box' around the list of units is missing a bottom line:

(minor issue, OK to go live if this issue is still there)

The Box around the list of units is missing a bottom line
![image](https://user-images.githubusercontent.com/31331637/45861368-c5740e80-bd9e-11e8-97cb-53690795eb32.png)

nbiton commented:
> I am still figuring out how to use react DOM to debug currently rendered RouteComp?

I think it's still easier to do by tracking the invocation of `componentDidMount` with console.log or breakpoints, anyway.

nbiton commented:
If it is so much trouble, I suggest we just pass on having this modal bound to the URL on all the screens it is used. I don't see that big of a benefit for it being URL-bound for it to be worth it spending so much time / adding complexity

sophieyang0813 commented:
@nbiton Good afternoon! 

I am mimicking the route flow of a `case item (/case/:caseId)` for the `unit-select` 
1) dispatch to storeBreadcrumb(/case) and have the route taken care of at /case, instead of direct dispatch to unit-select (/case/unit-select) 
2) put `unit select` render details in case.jsx, instead of indicating directly below `add` unit select button in case-explorer.jsx
 
I am still trying to figure out why the unit-select page wouldn't render.

sophieyang0813 commented:
@nbiton  I will try componentDidMount again after lunch. If I cannot figure out in an hour, then I will just revert back to non-URL bound modal design. 🌻🌻

franck-boullier commented:
BTW @sophieyang0813 this fix by @nbiton to solve button that was not working in FF might give you some ideas on how to solve the FF browser compatibility issue we have in this PR...

franck-boullier commented:
Status Update:

# Fixed issues:

- On the 'Case' page, endless load when clicking on the '+' button -  FIXED

# Pending issues:

## Unable to create a new case:

- Go to case
- Click on the '+' button
- Select a unit to create a new case
- Click on 'Create Case'

## Expected result:

User is sent to the 'Case creation' page for this unit

## Actual result:

Endless page load
url is .../case/new?unit=xxx

(Where xxx is the unit id of the unit which what selected).

## Browser compatibility (FF) - Impossible to create a Report (or a case):

Test on FF on a computer.
The problem does NOT happen on Chrome on Android (Galaxy S8)

- Go to the report (or case) page
- click on the '+' icon to add a report
- Select a unit
- Click on 'Create Report'

### Expected result:

User is sent to the "name your report" step.

### Actual result:

Nothing happens, the window stays open
![image](https://user-images.githubusercontent.com/31331637/45985523-37976c80-c099-11e8-8d4a-e320a83cf0e3.png)

## 'Square box' around the list of units is missing a bottom line:

(minor issue, OK to go live if this issue is still there)

The Box around the list of units is missing a bottom line
![image](https://user-images.githubusercontent.com/31331637/45861368-c5740e80-bd9e-11e8-97cb-53690795eb32.png)

sophieyang0813 commented:
@nbiton Evening! this code is ready for review (finally... 🌻). Modals work as intended, though they are not URL bound. 

Why modal without URL? 
URL bound unit-select-dialog worked in the case page (Thanks to your routing help). But, the case creation page raised API error.  FireFox complicated my solution to hiding modal on dismissed in the report page. I used Link's state property to discern whether to `go back` or `/report` page. But FireFox was not friendly to `Link` accompanied by `Raised button`. 

@franck-boullier all the issues are addressed except the square box. 
-> This customization would take a while...Material UI 's `autoScrollBodyContent` property
draws the borderline for scrollbar around the body, not around the unit list.

franck-boullier commented:
Status Update - Almost there!!!

# Fixed issues:

- The modal does not work for new case creation -  FIXED
- Browser compatibility (FF) - Impossible to create a Report (or a case) - FIXED

# Suggested improvement to save one click:

We can save one click in the Unit selection process (and we should) by slightly changing the flow for unit selection.

## The current flow:

To create a report or a case we need to:
- Click on the '+' icon
- Select a unit
- Click on 'Create Case' or 'Create Report'
---> Go to the create case page or go to the modal to name the report.

## Suggestion for a simplified flow:

We can save one click if we remove the 'Create Case'/'Create Report' button completely.
The simplified flow would be:
- Click on the '+' icon
- Select a unit
---> Go to the create case page or go to the modal to name the report.

We need to keep the 'Add a new unit' option of course in case the unit does not exist.

@sophieyang0813 do you see this as complex to implement?


# Minor issues (OK to fix in a different PR)

- 'Square box' around the list of units is missing a bottom line

franck-boullier commented:
Another minor thing:
Modal title is "Select units" it should be "Select Unit" (capital letter for consistency and singular form as we allow user to select only one unit in that window.

franck-boullier commented:
Status Update:

# Fixed Issue:

- Save one click on creating  a case or a report - FIXED

# Pending issue:

## Unit name "bleeds out" to the right of the screen:

![image](https://user-images.githubusercontent.com/31331637/46056049-65f07700-c181-11e8-80f2-4f22cbd3840f.png)

# Minor issues (OK to fix in a different PR)

- 'Square box' around the list of units is missing a bottom line

franck-boullier commented:
Status Update: all work as intended 🎉 Well done @sophieyang0813 !!

@nbiton I let you review this code and then we can push that in prod ASAP

nbiton commented on imports/ui/dialogs/unit-select-dialog.jsx:
```diff
@@ -0,0 +1,148 @@
+    const title =
+      <div>
+        <div className='pb1'>Select Unit</div>
+        <span className='ba b--moon-gray mv1 br1 pv1'>
```
This should be taken out of the title. The title should be just the title, and the rest of the elements should be a part of the body.

nbiton commented:
Issues I found so far:

### The search and results boxes are misaligned
![screenshot from 2018-09-26 13-02-32](https://user-images.githubusercontent.com/2662819/46058396-7b6b9e00-c18d-11e8-96ff-ef9fdad8c907.png)
(Red lines were added for highlighting it)
*Suggested fix*: See my comment on the code above

### Unit items design doesn't match the intended design
![screenshot from 2018-09-26 13-11-46](https://user-images.githubusercontent.com/2662819/46058685-a9051700-c18e-11e8-8da8-0d49f49c158f.png)
* Unit is only one line, doesn't include the address like in the design
* The title isn't centered between the separator lines
* The horizontal distance of the separators from the frame (the gray box) is too big compared to the design

### The "X" button is not aligned with "Select Unit"

### Scrolling  is done outside of the results' gray box instead of inside, which makes the top and bottom borders disappear while scrolling
![screenshot from 2018-09-26 13-23-06](https://user-images.githubusercontent.com/2662819/46058885-61cb5600-c18f-11e8-994c-d193ecec3f5e.png)

### The dialog's height jumps in a glitchy way when the number of results is smaller than the scrollable height
![screenshot from 2018-09-26 13-24-46](https://user-images.githubusercontent.com/2662819/46058952-b969c180-c18f-11e8-9fe5-cf7a3f732fc5.png)
*Suggested fix*: give the scrollable area a fixed/stretched (`flex-grow`) height

### No results looks like this
![screenshot from 2018-09-26 13-27-06](https://user-images.githubusercontent.com/2662819/46059003-efa74100-c18f-11e8-945e-0b48121ebd2a.png)
*Suggested fix*: Add a message that says no results were found. (Might require @kiatlim or @michkwt 's input)

### If the dialog is opened shortly after the route is navigated to, the results list is empty and fills gradually during loading
*Suggested fix*: show a preloader animation until all the unit names are loaded

franck-boullier commented:
> Unit is only one line, doesn't include the address like in the design

@nbiton this is OK ---> much cleaner like this IMO and in this context, we do not need to display the unit description (the design in Figma has not yet been updated to reflect that decision)

sophieyang0813 commented:
@nbiton Good evening! Thank you for feedback. I addressed most issues. Some remaining questions are following. 

> Scrolling is done outside of the results' gray box instead of inside

@franck-boullier also raised this issue. How did you customize a scrollable box for potential invitees in InviteDialog? I read  selectControlsRenderers and got lost a bit... 😅This will be helpful to figure out flex height as well. Somehow, Material UI is resistant to flex-grow for height when I tried. 

> The dialog's height jumps in a glitchy way when the number of results is smaller than the scrollable height

I chose min-height after several failed attempts using flex for height. 

> The horizontal distance of the separators from the frame (the gray box) is too big compared to the design

I didn't fully understood what this meant. But I did fix unit item. Could you see whether this has been fixed? 

I wll be emceeing for PhP conference tomorrow and Friday. But I will work on this during my spare time! I would love to close this PR... 🌻🌻

nbiton commented on imports/ui/dialogs/unit-select-dialog.jsx:
```diff
@@ -0,0 +1,160 @@
+          </span>
+        </div>
+        <div className='ba b--moon-gray pa2 br1' style={{minHeight: '380px'}}>
+          { isLoading ? (<Preloader />
```
Use `<CircularProgress ... />` instead. The `<Preloader` is only meant to be used as a full page preloader as it takes up the entire height of the screen and will not be centered in the container

nbiton commented on imports/ui/dialogs/unit-select-dialog.jsx:
```diff
@@ -0,0 +1,160 @@
+                    <div key={name}>
+                      <MenuItem
+                        innerDivStyle={resetMenuItemDivStyle}
+                        onClick={() => item === 'Report'
```
This should not be report/case specific. Suggested solution: let each user of this modal send in a function as prop to resolve the "next step path"- something of the form `(unitId) => '/something/action/${unitId}/item'` (just an example)

nbiton commented on imports/ui/dialogs/unit-select-dialog.jsx:
```diff
@@ -0,0 +1,160 @@
+                          ) : (
+                            dispatch(push(`/case/new?unit=${id}`))
+                          )}
+                        style={{whiteSpace: 'none', display: 'inline'}}
```
This seems redundant. What are these styles used for?

nbiton commented on imports/ui/dialogs/unit-select-dialog.jsx:
```diff
@@ -0,0 +1,160 @@
+    const actions =
+      <div>
+        <div className={'bondi-blue tc pa2'} onClick={() => dispatch(push(`/unit/new`))} >
+          <span className={units.length === 0 && 'dn'}> Add a new unit </span>
```
Space should not be used as padding. You can use `ph2` as a class name instead

nbiton commented on imports/ui/explorer-components/no-item-msg.jsx:
```diff
@@ -21,7 +22,15 @@ export function NoItemMsg ({item}) {
           <div className='mid-gray b lh-copy'>
             <div>There are no {item}s that match your current filter combination.</div>
-            <div>Click on the "+" button to select a unit to add a new {item}</div>
+            {item === 'unit' ? (
```
This component shouldn't be unit specific. It should be enhanced to facilitate what the unit select needs without mentioning 'unit' specifically

nbiton commented on imports/ui/unit-explorer/unit-type-icon.jsx:
```diff
@@ -52,5 +52,6 @@ export default class UnitTypeIcon extends Component {
   metaData: PropTypes.object,
-  iconInExplorer: PropTypes.string
+  iconInExplorer: PropTypes.string,
+  unitType: PropTypes.string
```
I can't find any usages of this new prop. Please remove it if it's not going to be used

nbiton commented on imports/ui/case/case-master.jsx:
```diff
@@ -39,6 +40,11 @@ class CaseMaster extends Component {
         RouteComp: CaseExplorer,
         exact: true
       },
+      // {
```
Remove commented out code

franck-boullier commented:
# Pending Issue:

Regression: scroll is not possible anymore

![image](https://user-images.githubusercontent.com/31331637/46182549-2bace400-c2ff-11e8-8304-241c8c730706.png)

franck-boullier commented:
Status Update:

# Fixed issues:

- The title isn't centered between the separator lines - FIXED
- The "X" button is not aligned with "Select Unit" - FIXED
- Add a message that says no results were found. - FIXED
- The dialog's title is misaligned horizontally with the rest of the content - FIXED
- The right border of the search box is misaligned with the results box - FIXED
- There's vertical padding to the results box which is unnecessary, and causing the first and last item to be rendered with a separator which shouldn't appear - FIXED

# Unable to test:

- The dialog's height jumps in a glitchy way when the number of results is smaller than the scrollable height - Suggested fix: give the scrollable area a fixed/stretched (flex-grow) height

# Pending Issue:

## Results list is empty and fills gradually during loading

If the dialog is opened shortly after the route is navigated to, the results list is empty and fills gradually during loading.
Suggested fix: show a preloader animation until all the unit names are loaded

## Regression: scroll is not possible anymore

![image](https://user-images.githubusercontent.com/31331637/46182549-2bace400-c2ff-11e8-8304-241c8c730706.png)

## (Minor) 'Add Unit' does not look like a button:

![image](https://user-images.githubusercontent.com/31331637/46182919-ee495600-c300-11e8-8d17-7217955040b9.png)


![image](https://user-images.githubusercontent.com/31331637/46182846-914da000-c300-11e8-82eb-2548a5837b59.png)

franck-boullier commented:
Also it might be good to reduce the top/bottom margin/padding in the list ---> this feels like unnecessary use of space:

![image](https://user-images.githubusercontent.com/31331637/46182964-34061e80-c301-11e8-8f44-98f208a3fbf0.png)

This looked better IMO

![image](https://user-images.githubusercontent.com/31331637/46182980-4e3ffc80-c301-11e8-9836-5ae35529faad.png)

franck-boullier commented:
Status Update:

# Fixed issues:

- 'Add Unit' does not look like a button - FIXED
- Smaller Top/Bottom padding in the Unit list - FIXED

# Unable to test:

- The dialog's height jumps in a glitchy way when the number of results is smaller than the scrollable height - Suggested fix: give the scrollable area a fixed/stretched (flex-grow) height
- Scrolling results in the search box being pushed out of the view, while the search icon remains. (Desired outcome: only the results are scrollable and the search box stays in place always)

![screenshot from 2018-09-26 21-18-20](https://user-images.githubusercontent.com/2662819/46082488-cad3bd80-c1d1-11e8-9131-d32486853380.png)

# Pending Issue:

## Results list is empty and fills gradually during loading

If the dialog is opened shortly after the route is navigated to, the results list is empty and fills gradually during loading.
Suggested fix: show a preloader animation until all the unit names are loaded

## Regression: scroll is not possible anymore

![image](https://user-images.githubusercontent.com/31331637/46182549-2bace400-c2ff-11e8-8304-241c8c730706.png)

franck-boullier commented:
Status Update

# Fixed Issue:

- Scroll in not possible in FF - FIXED

# Pending Issues:

# Pending Issue:

## Results list is empty and fills gradually during loading

If the dialog is opened shortly after the route is navigated to, the results list is empty and fills gradually during loading.
Suggested fix: show a preloader animation until all the unit names are loaded

## The modal is not displaying all search results:

Test done on a Samsung Galaxy S8 (Chrome)

- Logged in as Leonel
- Open modal to select a unit
- type a search text (ex 'Test')

### Expected result:

A scrollable list of unit matching the search term is visible

### actual result:

- only 1 line of text is visible
- list in not scrollable

## The modal is not centered anymore after search result are displayed:

Test done on a Samsung Galaxy S8 (Chrome)

- Logged in as Leonel
- Open modal to select a unit
- type a search text (ex 'Test')
- close keyboard
- Modal is all over the place...

## FF Browser compatibility issue - Scroll

The last value in the list is not visible when scrolling all the way down 

![image](https://user-images.githubusercontent.com/31331637/46189867-6d4f8600-c323-11e8-9fa5-a0870e83143c.png)

franck-boullier commented:
Status Update

# Fixed Issue:

- Results list is empty and fills gradually during loading - FIXED

# Pending Issues:

These are mostly browser compatibility issues.
@nbiton can you review this and see if we can merge already?

## Browser compatibility - The modal is not displaying all search results:

This seems to work on iOS/iPhone
Test done on a Samsung Galaxy S8 (Chrome)

- Logged in as Leonel
- Open modal to select a unit
- type a search text (ex 'Test')

### Expected result:

A scrollable list of unit matching the search term is visible

### actual result:

- only 1 line of text is visible
- list in not scrollable

## Browser compatibility - The modal is not centered anymore after search result are displayed:

This seems to work on iOS/iPhone
Test done on a Samsung Galaxy S8 (Chrome)

- Logged in as Leonel
- Open modal to select a unit
- type a search text (ex 'Test')
- close keyboard
- Modal is all over the place...

## FF Browser compatibility issue - Scroll

The last value in the list is not visible when scrolling all the way down 

![image](https://user-images.githubusercontent.com/31331637/46189867-6d4f8600-c323-11e8-9fa5-a0870e83143c.png)

sophieyang0813 commented:
@nbiton. This code is ready for review. 

* The button and unit margin have been fixed as per Franck's request. 
* loading icon shows before the unit list is populated. 
* height is added since the overflow-auto would not show the scrollbar

*somehow, my FireFox and Chrome show the last item in the unit list, while Franck's does not. 

**My Firefox** ( market street == last item) 
<img width="642" alt="screen shot 2018-09-28 at 1 50 29 pm" src="https://user-images.githubusercontent.com/25441733/46191024-7b53d580-c328-11e8-9579-7d96eb16f78f.png">

nbiton commented:
@sophieyang0813 There should be a consistent 16px gap from the edge of the dialog window to its content on all directions. Currently it's 12px at the top, 24 on the sides and 8 at the bottom.
I think there should be a bigger gap between "Add a new unit" button to the results box as well, although it's not reflected in the design as the button isn't the same.

nbiton commented:
Positioning issue:
![screenshot_20180928-153039_chrome](https://user-images.githubusercontent.com/2662819/46194327-b4de0e00-c333-11e8-92a9-6bc8b3d08cc6.jpg)
On Andorid (didn't test iOS), if the keyboard shows up and then closed, the bottom part of the dialog is pushed down, so the the button is inaccessible. Only way to get it back is to close the dialog and open it again.
This is not critical though

nbiton commented:
Scrolling issue:
![screenshot_20180928-153743_chrome](https://user-images.githubusercontent.com/2662819/46194821-15217f80-c335-11e8-8a55-413514dbe81d.jpg)
![screenshot_20180928-153756_chrome](https://user-images.githubusercontent.com/2662819/46194822-15ba1600-c335-11e8-9897-c7e8b9bbfb78.jpg)

On Android (didn't test iOS), when the keyboard is open the bottom part of the results list is cropped off, so 5 and a half search results are hidden and inaccessible.

This is more critical than the last issue I spotted

sophieyang0813 commented:
@nbiton Good evening! I just finished. This code is ready for review.

franck-boullier commented:
As far as I can see all the problems are now fixed.
@nbiton is it OK to merge now 🤞 ?

nbiton commented on imports/ui/dialogs/unit-select-dialog.jsx:
```diff
@@ -0,0 +1,155 @@
+            />
+          </div>
+        </div>
+        <div className='flex flex-column flex-grow overflow-auto'>
```
Please combine this div with the one below it (remove one and have both of their classNames on one dib) so the scrolling is done inside the grey box and not out of it

nbiton commented on imports/ui/explorer-components/no-item-msg.jsx:
```diff
@@ -2,7 +2,7 @@ import React from 'react'
 import PropTypes from 'prop-types'
 
-export function NoItemMsg ({item}) {
+export function NoItemMsg ({item, ButtonOption}) {
```
Add `buttonOption` (camelCase) to prop types

nbiton commented:
@sophieyang0813 good job on making the fixes, it's almost ready, just a few small final fixes.

@franck-boullier I only tested it functionally briefly, but I think it looks good too. Should be mergeable after this.

sophieyang0813 commented:
@nbiton Thank you for the review. Hope that it is mergeable now 😄😄!

nbiton commented on imports/ui/dialogs/unit-select-dialog.jsx:
```diff
@@ -0,0 +1,153 @@
+            </div>
+          ) : (
+            !isLoading ? (
+              <NoItemMsg item='unit' buttonOption={false} />
```
Setting `buttonOption={false}` isn't necessary. Just don't mention it at all to achieve the same effect

nbiton commented on imports/ui/case-explorer/case-explorer.jsx:
```diff
@@ -152,13 +159,18 @@ class CaseExplorer extends Component {
               }
               name={'case'}
-            /> : (<NoItemMsg item={'case'} />)
+            /> : (<NoItemMsg item={'case'} ButtonOption />)
```
This doesn't work because you didn't rename it when you renamed the component's prop to `buttonOption`

nbiton commented on imports/ui/report-explorer/report-explorer.jsx:
```diff
@@ -87,13 +99,19 @@ class ReportExplorer extends Component {
                 }
                 name={'report'}
-              /> : (<NoItemMsg item={'report'} />)
+              /> : (<NoItemMsg item={'report'} ButtonOption />)
```
This doesn't work because you didn't rename it when you renamed the component's prop to `buttonOption`

nbiton commented on imports/ui/report-explorer/report-explorer.jsx:
```diff
@@ -87,13 +99,19 @@ class ReportExplorer extends Component {
+            <UnitSelectDialog
+              show={open}
+              onDismissed={() => this.setState({open: false})}
+              item='Report'
```
The prop item is not in use anymore. Please remove it

sebdeckers commented:
🌟 Congrats! 👏🏻
